contracts/smart-account/modules/SessionValidationModules/ABISessionValidationModule.sol

```diff
@@ -0,0 +1,164 @@
+            ECDSA.recover(
+                ECDSA.toEthSignedMessageHash(_userOpHash),
+                _sessionKeySignature
+            );
+    }
+
+    /**
+     * @dev validates that the call (destinationContract, callValue, funcCallData)
+     * complies with the Session Key permissions represented by sessionKeyData
+     * @param destinationContract address of the contract to be called
+     * @param callValue value to be sent with the call
+     * @param _funcCallData the data for the call. is parsed inside the SVM
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     * param _callSpecificData additional data, for example some proofs if the SVM utilizes merkle trees itself
+     * for example to store a list of allowed tokens or receivers
+     */
+    function validateSessionParams(
+        address destinationContract,
+        uint256 callValue,
+        bytes calldata _funcCallData,
+        bytes calldata _sessionKeyData,
+        bytes memory /*_callSpecificData*/
+    ) public pure virtual override returns (address) {
+        (address sessionKey, Permission memory permission) = abi.decode(
+            _sessionKeyData,
+            (address, Permission)
+        );
+
+        if (destinationContract != permission.destinationContract) {
+            revert("ABISV Wrong Destination");
+        }
+        if (callValue > permission.valueLimit) {
+            revert("ABISV Value exceeded");
+        }
+
+        if (!checkPermission(_funcCallData, permission))
+            revert("ABISV: Permission violated");
+
+        return sessionKey;
+    }
+
+    function checkPermission(
+        bytes calldata data,
+        Permission memory permission
+    ) internal pure returns (bool) {
+        if (bytes4(data[0:4]) != permission.selector) return false;
+        uint256 length = permission.rules.length;
+        for (uint256 i; i < length; ) {
+            Rule memory rule = permission.rules[i];
+            bytes32 param = bytes32(data[4 + rule.offset:4 + rule.offset + 32]);
+            if (rule.condition == Condition.EQUAL && param != rule.value) {
+                return false;
+            } else if (
+                rule.condition == Condition.LESS_THAN_OR_EQUAL &&
+                param > rule.value
+            ) {
+                return false;
+            } else if (
+                rule.condition == Condition.LESS_THAN && param >= rule.value
+            ) {
+                return false;
+            } else if (
+                rule.condition == Condition.GREATER_THAN_OR_EQUAL &&
+                param < rule.value
+            ) {
+                return false;
+            } else if (
+                rule.condition == Condition.GREATER_THAN && param <= rule.value
+            ) {
+                return false;
+            } else if (
+                rule.condition == Condition.NOT_EQUAL && param == rule.value
+            ) {
+                return false;
+            }
```

Comment:
Writing it like this saves ~150 gas. Not sure if worth it
```
bool rulePassed = (ruleCondition == Condition.EQUAL &&
    param == ruleValue) ||
    (ruleCondition == Condition.LESS_THAN_OR_EQUAL &&
        param <= ruleValue) ||
    (ruleCondition == Condition.LESS_THAN && param < ruleValue) ||
    (ruleCondition == Condition.GREATER_THAN_OR_EQUAL &&
        param >= ruleValue) ||
    (ruleCondition == Condition.GREATER_THAN &&
        param > ruleValue) ||
    (ruleCondition == Condition.NOT_EQUAL && param != ruleValue);

if (!rulePassed) {
    return false;
}
```

Reply:
Hmm, this saves ~400 gas. Slightly better?
```
  bool rulePassed;
  assembly ("memory-safe") {
      switch ruleCondition
      case 0 {
          // Condition.EQUAL
          rulePassed := eq(param, ruleValue)
      }
      case 1 {
          // Condition.LESS_THAN_OR_EQUAL
          rulePassed := or(lt(param, ruleValue), eq(param, ruleValue))
      }
      case 2 {
          // Condition.LESS_THAN
          rulePassed := lt(param, ruleValue)
      }
      case 3 {
          // Condition.GREATER_THAN_OR_EQUAL
          rulePassed := or(gt(param, ruleValue), eq(param, ruleValue))
      }
      case 4 {
          // Condition.GREATER_THAN
          rulePassed := gt(param, ruleValue)
      }
      case 5 {
          // Condition.NOT_EQUAL
          rulePassed := not(eq(param, ruleValue))
      }
  }

  if (!rulePassed) {
      return false;
  }
  ```

Reply:
It's strange that even this doesn't generate a JUMP Table. I would assume for such a simple switch case the compiler would compile it down a constant gas JUMP table:
<img width="1142" alt="image" src="https://github.com/bcnmy/scw-contracts/assets/16562513/b1629524-1771-4c59-8521-b3f840b4deaf">